rest-api/src/main/scala/za/co/absa/enceladus/rest_api/controllers/v3/VersionedModelControllerV3.scala

```diff
@@ -0,0 +1,218 @@
+
+abstract class VersionedModelControllerV3[C <: VersionedModel with Product
+  with Auditable[C]](versionedModelService: VersionedModelService[C]) extends BaseController {
+
+  import za.co.absa.enceladus.rest_api.utils.implicits._
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  // todo maybe offset/limit?
+  @GetMapping(Array(""))
+  @ResponseStatus(HttpStatus.OK)
+  def getList(@RequestParam searchQuery: Optional[String]): CompletableFuture[Seq[VersionedSummary]] = {
+    versionedModelService.getLatestVersionsSummarySearch(searchQuery.toScalaOption)
+  }
+
+  @GetMapping(Array("/{name}"))
+  @ResponseStatus(HttpStatus.OK)
+  def getVersionSummaryForEntity(@PathVariable name: String): CompletableFuture[VersionedSummary] = {
+    versionedModelService.getLatestVersionSummary(name) map {
+      case Some(entity) => entity
+      case None => throw notFound()
+    }
+  }
+
+  @GetMapping(Array("/{name}/{version}"))
+  @ResponseStatus(HttpStatus.OK)
+  def getVersionDetail(@PathVariable name: String,
+                       @PathVariable version: String): CompletableFuture[C] = {
+    forVersionExpression(name, version)(versionedModelService.getVersion).map {
+      case Some(entity) => entity
+      case None => throw notFound()
+    }
+  }
+
+
+  @GetMapping(Array("/{name}/audit-trail"))
+  @ResponseStatus(HttpStatus.OK)
+  def getAuditTrail(@PathVariable name: String): CompletableFuture[AuditTrail] = {
+    versionedModelService.getAuditTrail(name)
+  }
+
+  @GetMapping(Array("/{name}/{version}/used-in"))
+  @ResponseStatus(HttpStatus.OK)
+  def usedIn(@PathVariable name: String,
+             @PathVariable version: String): CompletableFuture[UsedIn] = {
+    forVersionExpression(name, version) { case (name, versionInt) => versionedModelService.getUsedIn(name, Some(versionInt)) }
+  }
+
+  @GetMapping(Array("/{name}/{version}/export"))
+  @ResponseStatus(HttpStatus.OK)
+  def exportSingleEntity(@PathVariable name: String, @PathVariable version: String): CompletableFuture[String] = {
+    forVersionExpression(name, version)(versionedModelService.exportSingleItem)
+  }
+
+  @GetMapping(Array("/{name}/export"))
+  @ResponseStatus(HttpStatus.OK)
+  def exportLatestEntity(@PathVariable name: String): CompletableFuture[String] = {
+    versionedModelService.exportLatestItem(name) // todo: remove in favor of the above? (that supports /{name}/latest/export)
+  }
+
+  @PostMapping(Array("/{name}/import"))
+  @ResponseStatus(HttpStatus.CREATED)
+  def importSingleEntity(@AuthenticationPrincipal principal: UserDetails,
+                         @PathVariable name: String,
+                         @RequestBody importObject: ExportableObject[C],
+                         request: HttpServletRequest): CompletableFuture[ResponseEntity[Validation]] = {
+    if (name != importObject.item.name) {
+      Future.failed(new IllegalArgumentException(s"URL and payload entity name mismatch: '$name' != '${importObject.item.name}'"))
```

Comment:
I am not sure I like these plaintext methods. I have a feeling 
`{ "message": "XYZ" }` would fit better.

Reply:
To further support my comment. 404 response is 
```json
{
	"timestamp": "2022-05-04T08:01:40.637+0000",
	"status": 404,
	"error": "Not Found",
	"message": "No message available",
	"path": "/menas/api-v3/mapping-tables/SourceSystemMappingTable/latest/default"
}
```

Reply:
@AdrianOlosutean @benedeki @Zejnilovic

Reply:
`RestResponse` or `RestError` application come in mind, perhaps

Reply:
Will be part of the next PR. Created issue #2062